### PR TITLE
Feature/news list screen

### PR DESCRIPTION
**Pull request by MathiasRossen**

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -90,6 +90,12 @@ dependencies {
     implementation("androidx.datastore:datastore-preferences:${Versions.DATASTORE_PREFERENCES_VERSION}")
     implementation("androidx.datastore:datastore-preferences-rxjava3:${Versions.DATASTORE_PREFERENCES_RXJAVA_VERSION}")
 
+    // Joda time
+    implementation("joda-time:joda-time:${Versions.JODA_TIME_VERSION}")
```
Comment (daimis) [inline app/build.gradle.kts]: Joda time is deprecated. Use java time

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.viewmodel.compose.viewModel
+import dk.mathiasrossen.onboardingapp.R
+import dk.mathiasrossen.onboardingapp.models.Article
+import dk.mathiasrossen.onboardingapp.ui.theme.ButtonColors
+import dk.mathiasrossen.onboardingapp.ui.theme.OnboardingAppTheme
+
+@Composable
+fun ArticleList(articleListViewModel: ArticleListViewModel = viewModel()) {
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt]: Duplicate name `ArticleList`. Name it as `ArticleScreen`

Also, you can do it like this: 
`fun ArticleList(articleListViewModel: ArticleListViewModel = hiltViewModel())`

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/SortButton.kt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package dk.mathiasrossen.onboardingapp.ui.articles
+
+import androidx.annotation.StringRes
+
+data class SortButton(@StringRes val titleResource: Int, val sortState: SortState, val onClick: () -> Unit)
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/SortButton.kt]: Never use higher-order functions﻿ inside the data class (`val onClick: () -> Unit`). You are making it very complicated. It is hard to follow. Can have huge side effects because you never know where the callback might land...

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+) {
+    val sortButtonItems = listOf(
+        SortButton(R.string.article_list_button_sort_popular_today, SortState.POPULAR_TODAY, onPopularTodayClick),
+        SortButton(
+            R.string.article_list_button_sort_popular_all_time,
+            SortState.POPULAR_ALL_TIME,
+            onPopularAllTimeClick
+        ),
+        SortButton(R.string.article_list_button_sort_newest, SortState.NEWEST, onNewestClick)
+    )
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt]: You can simplify this by making 
 ```
 enum class SortState(@StringRes name: Int,) {
    POPULAR_TODAY(R.string.something),
```
When for items do it like this:
```
items(SortState.entries) {
```

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                    top = rowContentPadding,
+                    end = rowContentPadding
+                ),
+                horizontalArrangement = Arrangement.spacedBy(
+                    dimensionResource(id = R.dimen.base_arrangement_space_medium)
+                )
+            ) {
+                items(sortButtonItems) { sortButton ->
+                    val colors =
+                        if (sortButton.sortState == currentSortState) ButtonColors.defaultFilled() else ButtonColors.defaultTonal()
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleList.kt]: The soft limit for the line is 100 symbols. The hard limit is 120. Do not exceed it. Use brackets instead.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/sources/SourcesScreen.kt [hunk 6]
```diff
@@ -8,29 +9,37 @@ import androidx.compose.foundation.lazy.items
...
 import androidx.lifecycle.viewmodel.compose.viewModel
+import dk.mathiasrossen.onboardingapp.R
 import dk.mathiasrossen.onboardingapp.models.NewsSource
 import dk.mathiasrossen.onboardingapp.ui.theme.OnboardingAppTheme
 import dk.mathiasrossen.onboardingapp.ui.theme.Typography
 
 @Composable
-fun SourcesScreen(sourcesScreenViewModel: SourcesScreenViewModel = viewModel()) {
+fun SourcesScreen(
+    sourcesScreenViewModel: SourcesScreenViewModel = viewModel(),
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/sources/SourcesScreen.kt]: hiltViewModel

File: app/src/main/java/dk/mathiasrossen/onboardingapp/utils/LocalDateTimeAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package dk.mathiasrossen.onboardingapp.utils
+
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.ToJson
+import org.joda.time.LocalDateTime
+import org.joda.time.format.DateTimeFormat
+
+class LocalDateTimeAdapter {
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/utils/LocalDateTimeAdapter.kt:8]: Why it is an adapter? Try to avoid names which android uses for the framework.
Better name is formatter.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/models/Article.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import org.joda.time.LocalDateTime
+
+data class Article(
+    val source: Source,
+    val title: String,
+    val author: String,
+    val description: String?,
+    val url: String,
+    val urlToImage: String?,
+    val publishedAt: LocalDateTime,
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/models/Article.kt:13]: This class is the response to an article. Should not be modified.
We are doing it like this -> response is the response -> when mapper, to map data into required data for screen.

If you wonder why. It's because the same response can be reused multiple times and can have multiple mappers for different views.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    Row(
+        modifier = Modifier.fillMaxWidth(),
+        horizontalArrangement = Arrangement.SpaceBetween,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Text(text = authorAndDate, style = Typography.labelLarge, color = BlueGray)
+        Image(
+            modifier = Modifier
+                .width(24.dp)
+                .height(24.dp),
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt]: It can be just `.size(24.dp)`

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleListViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    init {
+        refresh()
+    }
+
+    fun refresh() {
+        isRefreshing.value = true
+
+        val sortBy =
+            if (sortState.value == SortState.POPULAR_TODAY || sortState.value == SortState.POPULAR_ALL_TIME) NewsApiService.SORT_BY_POPULAR else NewsApiService.SORT_BY_PUBLISHED_AT
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleListViewModel.kt]: Rearrange the code.

**Review by daimis — CHANGES_REQUESTED** [review]

File: app/build.gradle.kts (hunk 1 above)
Comment (MathiasRossen) [reply app/build.gradle.kts]: There are requirements for API 24 and java.time is supported from API 26

File: app/src/main/java/dk/mathiasrossen/onboardingapp/utils/LocalDateTimeAdapter.kt (hunk 7 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/utils/LocalDateTimeAdapter.kt:8]: Moshi documentation uses 'Adapter' for their examples:
https://github.com/square/moshi#custom-type-adapters

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                Text(
+                    text = article.description,
+                    style = Typography.bodyMedium,
+                    maxLines = 2,
+                    overflow = TextOverflow.Ellipsis
+                )
+            }
+        }
+        if (showDivider) {
+            Divider(modifier = Modifier.run { height(dimensionResource(id = R.dimen.base_divider_size)) })
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt]: Isn't the same as` Modifier.heiht`?

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    top = rowContentPadding,
+                    end = rowContentPadding
+                ),
+                horizontalArrangement = Arrangement.spacedBy(
+                    dimensionResource(id = R.dimen.base_arrangement_space_medium)
+                )
+            ) {
+                items(SortState.entries) { sortState ->
+                    val colors =
+                        if (sortState == currentSortState) {
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: If it's enough space `if` statement can be put on the same line.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                            ButtonColors.defaultFilled()
+                        } else {
+                            ButtonColors.defaultTonal()
+                        }
+                    val onButtonClick = when (sortState) {
+                        SortState.POPULAR_TODAY -> onPopularTodayClick
+                        SortState.POPULAR_ALL_TIME -> onPopularAllTimeClick
+                        else -> onNewestClick
+                    }
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: This should be inside view model.
Then it is possible to replace all 3 click events with one:
`onSortOptionClick: (sortState: SortState) -> Unit`

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        articleListViewModel::refresh,
+        articleListViewModel::sortByPopularToday,
+        articleListViewModel::sortByPopularAllTime,
+        articleListViewModel::sortByNewest
+    )
+}
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun ArticleList(
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: Same name `ArticleList`. Rename the first one to `ArticleScreen` and make the second one private.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleListViewModelTest.kt [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    fun sortByPopularToday_sortStateEqualsPopularToday() {
+        viewModel.sortByPopularToday()
+
+        Assert.assertEquals(SortState.POPULAR_TODAY, viewModel.sortState.value)
+    }
+
+    @Test
+    fun sortByPopularAll_sortStateEqualsPopularAll() {
+        given(service.getArticlesFromSource(eq(sourceId), any(), any(), eq(null))).willReturn(
```
Comment (daimis) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleListViewModelTest.kt]: Don't use `any()` in unit test. Use the data which is passed for this case.

Because it is easy to make this unit test irrelevant. For example, if you change the type of parameter this unit test will still pass but maybe it should fail.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/models/Article.kt (hunk 8 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/models/Article.kt:13]: I think it doesn't make sense to keep the date as a string.
Wouldn't it be okay to modify the response that fits the entire system better? Especially in a case where you figured out that the earlier iteration of the class was fundamentally wrong?

I'm trying to revamp the responses/internal classes in another branch at the moment to keep them separate. One for Retrofit and one for Room.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt (hunk 11 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticleItem.kt]: Yes.. I have no idea why I made this 😅

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt (hunk 12 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: Maybe we should make formatter rules, so the auto formatter can catch things like this?

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        articlesScreenViewModel.sortState.value,
+        articlesScreenViewModel.isRefreshing.value,
+        articlesScreenViewModel::refresh,
+        articlesScreenViewModel::setSortState
+    )
+}
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+private fun ArticlesScreen(
```
Comment (daimis) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: Duplicate name again. Now we have two `ArticlesScreen`

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+        listOf(
+            Article.createSample(),
+            Article.createSample()
+        )
+    )
+
+    @Before
+    fun setup() {
+        given(dateUtils.oneDayPast()).willReturn(LocalDate.parse(date))
+        given(service.getArticlesFromSource(eq(sourceId), eq(language), eq(sortBy), eq(date))).willReturn(
```
Comment (daimis) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt]: Most likely, argument matcher `eq()` is not required anymore.

**Review by daimis — APPROVED** [review]

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt (hunk 12 above)
Comment (daimis) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: I think it exists by default. Not the same but if it is more than 120 it will move the code to a new line. You just need to add in settings those limits.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt (hunk 16 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/ArticlesScreen.kt]: Is function overloading a problem? I feel this ArticleScreen is descriptive because that's what it actually is.
Here's an example from Androids own sunflower example: https://github.com/android/sunflower/blob/main/app/src/main/java/com/google/samples/apps/sunflower/compose/plantlist/PlantListScreen.kt

Ofcourse I will just change it if it fits the code style better. I just want to pick your thoughts on this :)

**Review by daimis — APPROVED** [review]
